### ISSUE TITLE
"window.exists is not a function" after loading another page in a script

### ISSUE THREAD

**Issue by laubsauger** [garris/BackstopJS#541]
Hi @garris,

i think there is a problem when a scenario script causes a new page to be loaded.

I'm trying to fill out a form on a page and submit the form afterwards (this works).
The form actions brings me to a new page where the screenshot is supposed to get taken.

As soon as the new page is loaded and backstop tries to capture the screenshot it throws an exception:
````bash
COMMAND | Command `reference` ended with an error after [5.807s]
COMMAND | BackstopException: Account-Debitorwahl on 320-smallestPhone: Error: An error has occurred evaluating the script in the browser.TypeError: window.exists is not a function
    at window._backstopSelectorsExpMap.window._backstopSelectorsExp.reduce (<anonymous>:13:28)
    at Array.reduce (<anonymous>)
    at result (<anonymous>:11:72)
    at <anonymous>:22:9
````
I had a quick look around and realised that this function is provided by `backstopTools.js` which is injected after the initial goToUrl via Chromy. They are never re-injected when loading another page.

I quickly tried to manually inject `backstopTools.js` via chromy after the page load and that fixes the issue.

See my onReadyScript for this scenario below

````javascript
const ConfigProvider = require('../../../_scripts/configProvider.js');
const path = require('path');

module.exports = (chromy, scenario, vp) => {
    const config = new ConfigProvider(scenario.url);

    chromy.type('input#email', config.getUsername());
    chromy.type('input#password', config.getPassword());
    chromy.click('form[action="/action/account/login"] input[type="submit"]', { waitLoadEvent: true });

    // Inject BackstopTools to provide helper methods needed for screenshots
    const BACKSTOP_TOOLS_PATH = path.join(__dirname, '../../../node_modules/backstopjs') + '/capture/backstopTools.js';
    chromy.inject('js', BACKSTOP_TOOLS_PATH);

    return chromy;
};
````

Backstop 3.0.19
Node 8.4.0
Chrome 61.0.3163.79 (Official Build) (64-bit)
CentOS7

**Comment by garris** [comment]
@laubsauger Thanks for posting this!   Yes.  This is something that I really want to fix.   I started reorganizing the code to make this happen -- but havent been able to finish it.

If you pull the new version you will see that file has been jammed into `runChromy.js`

https://github.com/garris/BackstopJS/blob/bc4dfc598bd3e9a40530713a845e820127bfdf5b/core/util/runChromy.js#L251-L258

Here the methods are all injected manually after page load. ^^^

But of course -- in your case -- loading another page happens AFTER this runs.   

I think the fix is to bundle all these injects into a single method and then listen for page load/unload events so any time a new URL is loaded, we can then immediately after, run the new single method -- so that calls to those methods always work.

Any help with this is greatly appreciated!

**Comment by laubsauger** [comment]
@garris that's what i've found and that's why i came to my hacky solution rather quickly :)

I've tried to setup a `ready` listener via `chromy.on("ready") ....` with my hack in a global `beforeScript` but this fired at rather odd moments and i was not able to get it to work properly. Didn't have much time to fiddle further though (i.e. look into the chromy code to see what is what).

**Comment by Dbuggerx** [comment]
Hey guys, I'm having the same issue as @laubsauger.
I believe the following _patch_ should fix the issue. @garris  could you please confirm?

```
diff --git a/core/util/runChromy.js b/core/util/runChromy.js
index ac6dded..94465fa 100644
--- a/core/util/runChromy.js
+++ b/core/util/runChromy.js
@@ -247,20 +247,7 @@ function processScenarioView (scenario, variantOrScenarioLabelSafe, scenarioLabe
   }
   chromy.goto(url);
 
-// =================UPDATE vvvv ==================
-  // --- load in backstopTools into client app ---
-  // chromy.inject('js', config.env.backstop + BACKSTOP_TOOLS_PATH);
-  chromy.defineFunction(exists);
-  chromy.defineFunction(isVisible);
-  chromy.defineFunction(expandSelectors);
-  chromy.defineFunction(startConsoleLogger);
-  chromy.defineFunction(hasLogged);
-  chromy.evaluate(_ => startConsoleLogger());
-  // console.log('Loading >>>' + config.env.backstop + BACKSTOP_TOOLS_PATH);
-  // chromy.inject('js', config.env.backstop + BACKSTOP_TOOLS_PATH);
-
-  // chromy.evaluate(() => console.log(exists))
-  // chromy.evaluate(() => console.log("window.exists ^^^^"))
+  defineBackstopFunctions(chromy);
 
 // =================UPDATE ^^^^^^^^^ ==================
 
@@ -347,6 +334,7 @@ function processScenarioView (scenario, variantOrScenarioLabelSafe, scenarioLabe
   }
 
   return new Promise((resolve, reject) => {
+    defineBackstopFunctions(chromy);
     chromy
       .evaluate(`window._selectorExpansion = '${scenario.selectorExpansion}'`)
       .evaluate(`window._backstopSelectors = '${scenario.selectors}'`)
@@ -386,6 +374,15 @@ function processScenarioView (scenario, variantOrScenarioLabelSafe, scenarioLabe
       .end()
       .catch(e => reject(new BackstopException('Chromy error', scenario, viewport, e)))
   });
+
+  function defineBackstopFunctions(chromy) {
+      chromy.defineFunction(exists);
+      chromy.defineFunction(isVisible);
+      chromy.defineFunction(expandSelectors);
+      chromy.defineFunction(startConsoleLogger);
+      chromy.defineFunction(hasLogged);
+      chromy.evaluate(_ => startConsoleLogger());
+  }
 }
 
 // vvv HELPERS vvv
```

**Comment by laubsauger** [comment]
@Dbuggerx at a quick glance i would say this does a bit of a cleanup and moves the "injection" to another place but it doesn't change the underlying issue. 

Your function `defineBackstopFunctions` will get executed only once per scenario (more or or less at the start of each scenario). If you happen do to a second page load (apart from the initial one), e.g. for following an internal link in a non-SPA these methods will still be `undefined` because they were not injected to the new page. 

That's why we need to hook into a page load event to know when to reinject.

I was playing with this in the morning and came up with a solution similar to yours with the added load event listener. Unfortunately i'm either missing something or it's just not my day but it's not working for me as of now.

I've put this before the  `return new Promise...` part and uncommented the `BACKSTOP_TOOLS_PATH` variable:

````javascript
chromy.on('Page.loadEventFired', () => {
      console.log('Page.loadEventFired');
      chromy.inject('js', config.env.backstop + BACKSTOP_TOOLS_PATH)
});
````

**Comment by Dbuggerx** [comment]
@laubsauger I've tried similar approaches as well, but those didn't work. So just to try a quick thing, I've changed the _runChromy.js_ file, and ran backstop again.
The errors are gone. Can you please try that?

**Comment by laubsauger** [comment]
@Dbuggerx Sorry for only judging by the looks of it. Went to try it just now and you're absolutely right.

script below executed as onReady script works now
````javscript
module.exports = (chromy, scenario, vp) => {
    chromy.goto('https://github.com/garris/BackstopJS/issues');
};
````

**Comment by Dbuggerx** [comment]
Np @laubsauger , I'm glad it works! 
Later on I'll try to send a PR (very busy right now).

**Comment by laubsauger** [comment]
I'm in the middle of preparing one, don't worry.

**Comment by garris** [comment]
Thank you both for looking into this.  Have you tried conditionally reinjecting these methods after the onReady script runs?  

I am on my phone so apologies if I misunderstood anything.

**Comment by laubsauger** [comment]
@garris I think there is no need for that. Even though i still don't get why xD

We will see. I've issued a PR.

**Comment by garris** [comment]
Ok great.  I will try to look at this today.

**Comment by garris** [comment]
I assume this was fixed by #550 
@laubsauger if you see any regressions please feel free to reopen.  Thank you very much for your help on this!

**Comment by laubsauger** [comment]
@garris You assume correctly :)

I've just tested 3.0.26 with my earlier hacks removed and it works flawlessly.
